src/main/java/com/suse/saltstack/netapi/datatypes/SaltVersion.java

```diff
@@ -0,0 +1,145 @@
+package com.suse.saltstack.netapi.datatypes;
+
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Represents a salt version
+ */
+public class SaltVersion implements Comparable<SaltVersion> {
+
+    private static Pattern SALT_VERSION_REGEX =
+            Pattern.compile("^(\\d+)\\.(\\d+)\\.(\\d+)([Rr][Cc](\\d+))?$");
+
+    private final int year;
+    private final int month;
+    private final int bugfix;
+    private final Optional<Integer> releaseCandidate;
+
+    /**
+     * Parses a salt version string
+     * @param versionString the salt version string
+     * @return SaltVersion if the versionString is valid or empty Optional if not.
+     */
+    public static Optional<SaltVersion> parse(String versionString) {
```

Comment:
The whole idea of `Optional` is to get rid of null and replace it with a safer alternative. What `Optional` does is it encodes the concept of something being present or not in the type itself. Similar to a `List` that encodes the possibility of multiple values being there. Think of it as List is a way to say there are 0 or more values while `Optional` is 0 or 1 value. The advantage of encoding things in the type is that the developer immediately sees whats going on without reading the docs. In contrast to null you have to actively deal with `Optional` by calling `map` , `ifPresent`, `orElse` or `getOrElse` (`get` excluded because its not safe and should not be used) otherwise you will not get access to its inner value if there is one. With those helper methods on `Optional` you can write your logic as if the value is there and `Optional` will do the "null checking" for you.